### ISSUE TITLE
Command line with no GUI

### ISSUE THREAD

**Issue by n2cr** [DSheirer/sdrtrunk#92]
On an Odroid-C2 embedded ARM computer [http://www.hardkernel.com/main/products/prdt_info.php](url) the GUI takes too much CPU leaving it unable to decode the P25 Phase 1 system I stream to my local network.

Would it be possible to have a GUI-free command line option which would allow streaming?

Streaming working great on a virtual hosted on my local server!  Thanks so much for this feature.  Better than OP25 with GNURadio.

**Comment by DSheirer** [comment]
This would be blocked on issues #34 and #53, both of which should improve the processing efficiency.  Both of these are the next big issues I'm going to tackle after I finish a few small bug issues.

**Comment by ImagoTrigger** [comment]
playing around with JAVA_TOOL_OPTIONS: -Djava.awt.headless=true

i run multiple copies of sdrtrunk on the same server (using my dirty hacks) each instance uses 10-15% of my GPUs ;-)

**Comment by chudgoo** [comment]
@ImagoTrigger  -  Is this running on an ODROID C2 or another aarch64 system?

**Comment by ImagoTrigger** [comment]
nah, amd64 windhoes

**Comment by n2cr** [comment]
Where does the JAVA_TOOL_OPTIONS go?

**Comment by ImagoTrigger** [comment]
R:\scanner\out\production\sdrtrunk>java **-Djava.awt.headless=true** -XX:+UseG1GC -cp "SDRTrunk.jar;*;libs/*;config/*;images/*;imports/*;" gui.SDRTrunk

results in less resource usage. 0% GPU (tested on windows)

because of my poor java skills i just duped SDRTrunk class to make mMainGui a Jpanel instead of a JFrame when headless is defined.  of course you could reorganize to make it less fugly. also if headless dont do any JideTabbedPane stuff...  but this WORKS!!!! now my GPU does not heat up running SDRTrunk and now I can run it on a linux VPS....

`diff --git a/src/alias/AliasEditor.java b/src/alias/AliasEditor.java
index 706ef5e..2ee8276 100644
--- a/src/alias/AliasEditor.java
+++ b/src/alias/AliasEditor.java
@@ -93,15 +93,17 @@ public class AliasEditor extends Editor<Alias>
 		add( mAliasLabel, "align left" );
 		
 		add( new JSeparator( JSeparator.HORIZONTAL ), "span,growx,push" );
-		
-		JideTabbedPane tabs = new JideTabbedPane();
-		tabs.setFont( this.getFont() );
-    	tabs.setForeground( Color.BLACK );
-
-		tabs.addTab( "Alias", mAliasNameEditor );
-		tabs.addTab( "Audio / Identifier", mAliasIdentifierEditor );
-		tabs.addTab( "Action", mAliasActionEditor );
-		add( tabs, "span,grow" );
+
+		if (!java.awt.GraphicsEnvironment.isHeadless()) {
+			JideTabbedPane tabs = new JideTabbedPane();
+			tabs.setFont(this.getFont());
+			tabs.setForeground(Color.BLACK);
+
+			tabs.addTab("Alias", mAliasNameEditor);
+			tabs.addTab("Audio / Identifier", mAliasIdentifierEditor);
+			tabs.addTab("Action", mAliasActionEditor);
+			add(tabs, "span,grow");
+		}
 		
 		//Playlist management buttons
 		JButton btnSave = new JButton( "Save" );
diff --git a/src/controller/ControllerPanel.java b/src/controller/ControllerPanel.java
index 9b3f6cd..606168b 100644
--- a/src/controller/ControllerPanel.java
+++ b/src/controller/ControllerPanel.java
@@ -78,7 +78,7 @@ public class ControllerPanel extends JPanel
 
         mAudioPanel = new AudioPanel(iconManager, settingsManager, sourceManager, audioManager);
 
-        mChannelMetadataViewer = new ChannelMetadataViewer(channelProcessingManager, iconManager, settingsManager);
+        if (!java.awt.GraphicsEnvironment.isHeadless()) mChannelMetadataViewer = new ChannelMetadataViewer(channelProcessingManager, iconManager, settingsManager);
 
         mMapPanel = new MapPanel(mapService, iconManager, settingsManager);
 
@@ -98,21 +98,22 @@ public class ControllerPanel extends JPanel
         setLayout(new MigLayout("insets 0 0 0 0 ", "[grow,fill]", "[]0[grow,fill]0[]"));
 
         add(mAudioPanel, "wrap");
-
-        //Tabbed View - configuration, calls, messages, map
-        mTabbedPane = new JideTabbedPane();
-        mTabbedPane.setFont(this.getFont());
-        mTabbedPane.setForeground(Color.BLACK);
-        mTabbedPane.addTab("Now Playing", mChannelMetadataViewer);
-        mTabbedPane.addTab("Aliases", mAliasController);
-        mTabbedPane.addTab("Channels", mChannelController);
-        mTabbedPane.addTab("Map", mMapPanel);
-        mTabbedPane.addTab("Streaming", mBroadcastPanel);
-        mTabbedPane.addTab("Tuners", mTunerManagerPanel);
-
-        //Set preferred size to influence the split between these panels
-        mTabbedPane.setPreferredSize(new Dimension(880, 500));
-
-        add(mTabbedPane,"wrap");
+        if (!java.awt.GraphicsEnvironment.isHeadless()) {
+            //Tabbed View - configuration, calls, messages, map
+            mTabbedPane = new JideTabbedPane();
+             mTabbedPane.setFont(this.getFont());
+            mTabbedPane.setForeground(Color.BLACK);
+           mTabbedPane.addTab("Now Playing", mChannelMetadataViewer);
+             mTabbedPane.addTab("Aliases", mAliasController);
+             mTabbedPane.addTab("Channels", mChannelController);
+           mTabbedPane.addTab("Map", mMapPanel);
+            mTabbedPane.addTab("Streaming", mBroadcastPanel);
+            mTabbedPane.addTab("Tuners", mTunerManagerPanel);
+
+            //Set preferred size to influence the split between these panels
+             mTabbedPane.setPreferredSize(new Dimension(880, 500));
+
+           add(mTabbedPane,"wrap");
+        }
     }
 }
diff --git a/src/controller/channel/ChannelEditor.java b/src/controller/channel/ChannelEditor.java
index 46edcd7..7a7be6e 100644
--- a/src/controller/channel/ChannelEditor.java
+++ b/src/controller/channel/ChannelEditor.java
@@ -78,38 +78,47 @@ public class ChannelEditor extends Editor<Channel> implements ActionListener, Ch
 	private void init()
 	{
 		setLayout( new MigLayout( "fill,wrap 3", "[grow,fill][grow,fill][grow,fill]", "[grow,fill][]" ) );
-		
-		JideTabbedPane tabs = new JideTabbedPane();
-		tabs.setFont( this.getFont() );
-    	tabs.setForeground( Color.BLACK );
 
-    	tabs.setTabTrailingComponent( mChannelName );
-    	
+
 		mNameConfigurationEditor.setSaveRequestListener( this );
-    	tabs.addTab( "Name/Alias", mNameConfigurationEditor );
-    	
 		mSourceConfigurationEditor = new SourceConfigurationEditor( mSourceManager );
 		mSourceConfigurationEditor.setSaveRequestListener( this );
-		tabs.addTab( "Source", mSourceConfigurationEditor );
-		
 		mDecodeConfigurationEditor = new DecodeConfigurationEditor( mChannelMapModel );
 		mDecodeConfigurationEditor.setSaveRequestListener( this );
-		tabs.addTab( "Decoder", mDecodeConfigurationEditor );
-
 		mAuxDecodeConfigurationEditor = new AuxDecodeConfigurationEditor();
 		mAuxDecodeConfigurationEditor.setSaveRequestListener( this );
-		tabs.addTab( "Aux Decoders", mAuxDecodeConfigurationEditor );
-
 		mEventLogConfigurationEditor = new EventLogConfigurationEditor();
 		mEventLogConfigurationEditor.setSaveRequestListener( this );
-		tabs.addTab( "Logging", mEventLogConfigurationEditor );
-		
 		mRecordConfigurationEditor = new RecordConfigurationEditor();
 		mRecordConfigurationEditor.setSaveRequestListener( this );
-		tabs.addTab( "Recording", mRecordConfigurationEditor );
 
-		add( tabs, "span" );
+		if (!java.awt.GraphicsEnvironment.isHeadless()) {
+			JideTabbedPane tabs = new JideTabbedPane();
+			tabs.setFont(this.getFont());
+			tabs.setForeground(Color.BLACK);
+
+			tabs.setTabTrailingComponent(mChannelName);
+
+
+			tabs.addTab("Name/Alias", mNameConfigurationEditor);
+
 
+			tabs.addTab("Source", mSourceConfigurationEditor);
+
+
+			tabs.addTab("Decoder", mDecodeConfigurationEditor);
+
+
+			tabs.addTab("Aux Decoders", mAuxDecodeConfigurationEditor);
+
+
+			tabs.addTab("Logging", mEventLogConfigurationEditor);
+
+
+			tabs.addTab("Recording", mRecordConfigurationEditor);
+
+			add(tabs, "span");
+		}
 		mEnableButton.addActionListener( this );
 		mEnableButton.setEnabled( false );
 		mEnableButton.setToolTipText( "Start the currently selected channel running/decoding" );
diff --git a/src/gui/SDRTrunk.java b/src/gui/SDRTrunk.java
index 401067f..27fce4e 100644
--- a/src/gui/SDRTrunk.java
+++ b/src/gui/SDRTrunk.java
@@ -196,7 +199,11 @@ public class SDRTrunk implements Listener<TunerEvent>
      */
     public static void main(String[] args)
     {
-        new SDRTrunk();
+        if (java.awt.GraphicsEnvironment.isHeadless()) {
+            new SDRTrunk2();
+        } else {
+            new SDRTrunk();
+        }
     }
 
     /**
diff --git a/src/gui/SDRTrunk2.java b/src/gui/SDRTrunk2.java
new file mode 100644
index 0000000..3ec66c3
--- /dev/null
+++ b/src/gui/SDRTrunk2.java
@@ -0,0 +1,502 @@
+package gui;
+
+import alias.AliasModel;
+import alias.action.AliasActionManager;
+import audio.AudioManager;
+import audio.broadcast.BroadcastModel;
+import audio.broadcast.BroadcastStatusPanel;
+import com.jidesoft.swing.JideSplitPane;
+import controller.ControllerPanel;
+import controller.channel.ChannelModel;
+import controller.channel.ChannelProcessingManager;
+import controller.channel.ChannelSelectionManager;
+import controller.channel.map.ChannelMapModel;
+import icon.IconManager;
+import map.MapService;
+import module.log.EventLogManager;
+import net.miginfocom.swing.MigLayout;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import playlist.PlaylistManager;
+import properties.SystemProperties;
+import record.RecorderManager;
+import sample.Listener;
+import settings.SettingsManager;
+import source.SourceManager;
+import source.tuner.TunerEvent;
+import source.tuner.TunerModel;
+import source.tuner.TunerSpectralDisplayManager;
+import source.tuner.configuration.TunerConfigurationModel;
+import spectrum.SpectralDisplayPanel;
+import util.ThreadPool;
+import util.TimeStamp;
+
+import javax.imageio.ImageIO;
+import javax.swing.*;
+import java.awt.*;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.awt.event.KeyEvent;
+import java.awt.image.BufferedImage;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * Created by zerom on 2/28/2017.
+ */
+public class SDRTrunk2 implements Listener<TunerEvent>
+{
+    private final static Logger mLog = LoggerFactory.getLogger(SDRTrunk.class);
+    private static final String PROPERTY_BROADCAST_STATUS_VISIBLE = "main.broadcast.status.visible";
+    private boolean mBroadcastStatusVisible;
+
+    private IconManager mIconManager;
+    private BroadcastStatusPanel mBroadcastStatusPanel;
+    private BroadcastModel mBroadcastModel;
+    private ControllerPanel mControllerPanel;
+    private SettingsManager mSettingsManager;
+    private SpectralDisplayPanel mSpectralPanel;
+
+    private JPanel mMainGui = new JPanel();
+    private JideSplitPane mSplitPane;
+
+    private String mTitle;
+
+    public SDRTrunk2()
+    {
+        mLog.info("");
+        mLog.info("*******************************************************************");
+        mLog.info("**** sdrtrunk: a trunked radio and digital decoding application ***");
+        mLog.info("****  website: https://github.com/dsheirer/sdrtrunk             ***");
+        mLog.info("*******************************************************************");
+        mLog.info("");
+        mLog.info("Host CPU Cores: " + Runtime.getRuntime().availableProcessors());
+        mLog.info("Host Memory Total: " + Runtime.getRuntime().totalMemory());
+        mLog.info("Host Memory Max: " + Runtime.getRuntime().maxMemory());
+        mLog.info("Host Memory Free: " + Runtime.getRuntime().freeMemory());
+        //Setup the application home directory
+        Path home = getHomePath();
+
+        ThreadPool.logSettings();
+
+        mLog.info("Home path: " + home.toString());
+
+        //Load properties file
+        if(home != null)
+        {
+            loadProperties(home);
+        }
+
+        //Log current properties setting
+        SystemProperties.getInstance().logCurrentSettings();
+
+        TunerConfigurationModel tunerConfigurationModel = new TunerConfigurationModel();
+        TunerModel tunerModel = new TunerModel(tunerConfigurationModel);
+
+        mIconManager = new IconManager();
+
+        mSettingsManager = new SettingsManager(tunerConfigurationModel);
+
+        AliasModel aliasModel = new AliasModel();
+
+        ChannelModel channelModel = new ChannelModel();
+
+        ChannelMapModel channelMapModel = new ChannelMapModel();
+
+        EventLogManager eventLogManager = new EventLogManager();
+
+        RecorderManager recorderManager = new RecorderManager();
+
+        SourceManager sourceManager = new SourceManager(tunerModel, mSettingsManager);
+
+        ChannelProcessingManager channelProcessingManager = new ChannelProcessingManager(
+                channelModel, channelMapModel, aliasModel, eventLogManager, recorderManager, sourceManager);
+        channelProcessingManager.addAudioPacketListener(recorderManager);
+
+        channelModel.addListener(channelProcessingManager);
+
+        ChannelSelectionManager channelSelectionManager =
+                new ChannelSelectionManager(channelModel);
+        channelModel.addListener(channelSelectionManager);
+
+        AliasActionManager aliasActionManager = new AliasActionManager();
+        channelProcessingManager.addMessageListener(aliasActionManager);
+
+        AudioManager audioManager = new AudioManager(sourceManager.getMixerManager());
+        channelProcessingManager.addAudioPacketListener(audioManager);
+
+        mBroadcastModel = new BroadcastModel(mIconManager);
+
+        channelProcessingManager.addAudioPacketListener(mBroadcastModel);
+
+        MapService mapService = new MapService(mIconManager);
+        channelProcessingManager.addMessageListener(mapService);
+
+        mControllerPanel = new ControllerPanel(audioManager, aliasModel, mBroadcastModel,
+                channelModel, channelMapModel, channelProcessingManager, mIconManager,
+                mapService, mSettingsManager, sourceManager, tunerModel);
+
+        mSpectralPanel = new SpectralDisplayPanel(channelModel,
+                channelProcessingManager, mSettingsManager);
+
+        TunerSpectralDisplayManager tunerSpectralDisplayManager =
+                new TunerSpectralDisplayManager(mSpectralPanel,
+                        channelModel, channelProcessingManager, mSettingsManager);
+        tunerModel.addListener(tunerSpectralDisplayManager);
+        tunerModel.addListener(this);
+
+        PlaylistManager playlistManager = new PlaylistManager(aliasModel, mBroadcastModel, channelModel,
+                channelMapModel);
+
+        playlistManager.init();
+
+        mLog.info("starting main application headless (no gui)");
+
+        //Initialize the GUI
+        initGUI();
+
+        tunerModel.requestFirstTunerDisplay();
+
+        //Start the gui
+        EventQueue.invokeLater(new Runnable()
+        {
+            public void run()
+            {
+                try
+                {
+                    mMainGui.setVisible(true);
+                }
+                catch(Exception e)
+                {
+                    e.printStackTrace();
+                }
+            }
+        });
+    }
+
+    /**
+     * Launch the application.
+     */
+    public static void main(String[] args)
+    {
+        if (java.awt.GraphicsEnvironment.isHeadless()) {
+            new SDRTrunk2();
+        } else {
+            new SDRTrunk();
+        }
+    }
+
+    /**
+     * Initialize the contents of the frame.
+     */
+    private void initGUI()
+    {
+        mMainGui.setLayout(new MigLayout("insets 0 0 0 0 ", "[grow,fill]", "[grow,fill]"));
+
+        /**
+         * Setup main JFrame window
+         */
+        mTitle = SystemProperties.getInstance().getApplicationName();
+        // mMainGui.setTitle(mTitle);
+        mMainGui.setBounds(100, 100, 1280, 800);
+        //mMainGui.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+
+        //Set preferred sizes to influence the split
+        mSpectralPanel.setPreferredSize(new Dimension(1280, 300));
+        mControllerPanel.setPreferredSize(new Dimension(1280, 500));
+
+        mSplitPane = new JideSplitPane(JideSplitPane.VERTICAL_SPLIT);
+        mSplitPane.setDividerSize(5);
+        mSplitPane.add(mSpectralPanel);
+        mSplitPane.add(mControllerPanel);
+
+        mBroadcastStatusVisible = SystemProperties.getInstance().get(PROPERTY_BROADCAST_STATUS_VISIBLE, false);
+
+        //Show broadcast status panel when user requests - disabled by default
+        if(mBroadcastStatusVisible)
+        {
+            mSplitPane.add(getBroadcastStatusPanel());
+        }
+
+        mMainGui.add(mSplitPane, "cell 0 0,span,grow");
+
+        /**
+         * Menu items
+         */
+        JMenuBar menuBar = new JMenuBar();
+        //mMainGui.setJMenuBar(menuBar);
+
+        JMenu fileMenu = new JMenu("File");
+        menuBar.add(fileMenu);
+
+        JMenuItem logFilesMenu = new JMenuItem("Logs & Recordings");
+        logFilesMenu.addActionListener(new ActionListener()
+        {
+            @Override
+            public void actionPerformed(ActionEvent arg0)
+            {
+                try
+                {
+                    Desktop.getDesktop().open(getHomePath().toFile());
+                }
+                catch(Exception e)
+                {
+                    mLog.error("Couldn't open file explorer");
+
+                    JOptionPane.showMessageDialog(mMainGui,
+                            "Can't launch file explorer - files are located at: " +
+                                    getHomePath().toString(),
+                            "Can't launch file explorer",
+                            JOptionPane.ERROR_MESSAGE);
+                }
+            }
+        });
+        fileMenu.add(logFilesMenu);
+
+        JMenuItem settingsMenu = new JMenuItem("Icon Manager");
+        settingsMenu.addActionListener(new ActionListener()
+        {
+            @Override
+            public void actionPerformed(ActionEvent arg0)
+            {
+                mIconManager.showEditor(mMainGui);
+            }
+        });
+        fileMenu.add(settingsMenu);
+
+        fileMenu.add(new JSeparator());
+
+        JMenuItem exitMenu = new JMenuItem("Exit");
+        exitMenu.addActionListener(
+                new ActionListener()
+                {
+                    public void actionPerformed(ActionEvent event)
+                    {
+                        System.exit(0);
+                    }
+                }
+        );
+
+        fileMenu.add(exitMenu);
+
+        JMenu viewMenu = new JMenu("View");
+
+        viewMenu.add(new BroadcastStatusVisibleMenuItem(mControllerPanel));
+
+        menuBar.add(viewMenu);
+
+        JMenuItem screenCaptureItem = new JMenuItem("Screen Capture");
+
+        screenCaptureItem.setMnemonic(KeyEvent.VK_C);
+        screenCaptureItem.setAccelerator(
+                KeyStroke.getKeyStroke(KeyEvent.VK_C, ActionEvent.ALT_MASK));
+
+        screenCaptureItem.addActionListener(new ActionListener()
+        {
+            @Override
+            public void actionPerformed(ActionEvent arg0)
+            {
+                try
+                {
+                    Robot robot = new Robot();
+
+                    final BufferedImage image =
+                            robot.createScreenCapture(mMainGui.getBounds());
+
+                    SystemProperties props = SystemProperties.getInstance();
+
+                    Path capturePath = props.getApplicationFolder("screen_captures");
+
+                    if(!Files.exists(capturePath))
+                    {
+                        try
+                        {
+                            Files.createDirectory(capturePath);
+                        }
+                        catch(IOException e)
+                        {
+                            mLog.error("Couldn't create 'screen_captures' "
+                                    + "subdirectory in the " +
+                                    "SDRTrunk application directory", e);
+                        }
+                    }
+
+                    String filename = TimeStamp.getTimeStamp("_") +
+                            "_screen_capture.png";
+
+                    final Path captureFile = capturePath.resolve(filename);
+
+                    EventQueue.invokeLater(new Runnable()
+                    {
+                        @Override
+                        public void run()
+                        {
+                            try
+                            {
+                                ImageIO.write(image, "png",
+                                        captureFile.toFile());
+                            }
+                            catch(IOException e)
+                            {
+                                mLog.error("Couldn't write screen capture to "
+                                        + "file [" + captureFile.toString() + "]", e);
+                            }
+                        }
+                    });
+                }
+                catch(AWTException e)
+                {
+                    mLog.error("Exception while taking screen capture", e);
+                }
+            }
+        });
+
+        menuBar.add(screenCaptureItem);
+    }
+
+    /**
+     * Lazy constructor for broadcast status panel
+     */
+    private BroadcastStatusPanel getBroadcastStatusPanel()
+    {
+        if(mBroadcastStatusPanel == null)
+        {
+            mBroadcastStatusPanel = new BroadcastStatusPanel(mBroadcastModel);
+            mBroadcastStatusPanel.setPreferredSize(new Dimension(880, 70));
+            mBroadcastStatusPanel.getTable().setEnabled(false);
+        }
+
+        return mBroadcastStatusPanel;
+    }
+
+    /**
+     * Toggles visibility of the broadcast channels status panel at the bottom of the controller panel
+     */
+    private void toggleBroadcastStatusPanelVisibility()
+    {
+        mBroadcastStatusVisible = !mBroadcastStatusVisible;
+
+        EventQueue.invokeLater(new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                if(mBroadcastStatusVisible)
+                {
+                    mSplitPane.add(getBroadcastStatusPanel());
+                }
+                else
+                {
+                    mSplitPane.remove(getBroadcastStatusPanel());
+                }
+
+                mMainGui.revalidate();
+            }
+        });
+
+        SystemProperties.getInstance().set(PROPERTY_BROADCAST_STATUS_VISIBLE, mBroadcastStatusVisible);
+    }
+
+
+    /**
+     * Loads the application properties file from the user's home directory,
+     * creating the properties file for the first-time, if necessary
+     */
+    private void loadProperties(Path homePath)
+    {
+        Path propsPath = homePath.resolve("SDRTrunk.properties");
+
+        if(!Files.exists(propsPath))
+        {
+            try
+            {
+                mLog.info("SDRTrunk - creating application properties file [" +
+                        propsPath.toAbsolutePath() + "]");
+
+                Files.createFile(propsPath);
+            }
+            catch(IOException e)
+            {
+                mLog.error("SDRTrunk - couldn't create application properties "
+                        + "file [" + propsPath.toAbsolutePath(), e);
+            }
+        }
+
+        if(Files.exists(propsPath))
+        {
+            SystemProperties.getInstance().load(propsPath);
+        }
+        else
+        {
+            mLog.error("SDRTrunk - couldn't find or recreate the SDRTrunk " +
+                    "application properties file");
+        }
+    }
+
+    /**
+     * Gets (or creates) the SDRTRunk application home directory.
+     *
+     * Note: the user can change this setting to allow log files and other
+     * files to reside elsewhere on the file system.
+     */
+    private Path getHomePath()
+    {
+        Path homePath = FileSystems.getDefault()
+                .getPath(System.getProperty("user.home"), "SDRTrunk");
+
+        if(!Files.exists(homePath))
+        {
+            try
+            {
+                Files.createDirectory(homePath);
+
+                mLog.info("SDRTrunk - created application home directory [" +
+                        homePath.toString() + "]");
+            }
+            catch(Exception e)
+            {
+                homePath = null;
+
+                mLog.error("SDRTrunk: exception while creating SDRTrunk home " +
+                        "directory in the user's home directory", e);
+            }
+        }
+
+        return homePath;
+    }
+
+    @Override
+    public void receive(TunerEvent event)
+    {
+        if(event.getEvent() == TunerEvent.Event.REQUEST_MAIN_SPECTRAL_DISPLAY)
+        {
+            //mMainGui.setTitle(mTitle + " - " + event.getTuner().getName());
+        }
+    }
+
+    public class BroadcastStatusVisibleMenuItem extends JCheckBoxMenuItem
+    {
+        private ControllerPanel mControllerPanel;
+
+        public BroadcastStatusVisibleMenuItem(ControllerPanel controllerPanel)
+        {
+            super("Show Streaming Status");
+
+            mControllerPanel = controllerPanel;
+
+            setSelected(mBroadcastStatusPanel != null);
+
+            addActionListener(new ActionListener()
+            {
+                @Override
+                public void actionPerformed(ActionEvent e)
+                {
+                    toggleBroadcastStatusPanelVisibility();
+                    setSelected(mBroadcastStatusVisible);
+                }
+            });
+        }
+    }
+}
diff --git a/src/gui/control/JFrequencyControl.java b/src/gui/control/JFrequencyControl.java
index ca2053a..ed77b43 100644
--- a/src/gui/control/JFrequencyControl.java
+++ b/src/gui/control/JFrequencyControl.java
@@ -123,6 +123,7 @@ public class JFrequencyControl extends JPanel implements IFrequencyChangeProcess
         //Create image for cursor using empty array
         Image cursorImage = Toolkit.getDefaultToolkit().createImage(imageByte);
 
+        if (!java.awt.GraphicsEnvironment.isHeadless())
         mBlankCursor = Toolkit.getDefaultToolkit()
                 .createCustomCursor(cursorImage, new Point(0, 0), "cursor");
 
`

**Comment by davidmckenzie** [comment]
That looks awesome, very keen to try it!

**Comment by ImagoTrigger** [comment]
i was able to confirm this works on a very old server not running any kind of display manager.  I also commented out the default mixer stuff so java does not even try to load up any sound resources. instead of crapping out with the java.awt.HeadlessException it actually starts recording wavs!  this was on a early pentium 4 /w only 512mb ram.

**Comment by chudgoo** [comment]
Hey @ImagoTrigger 
Can you provide some more detail about how to implement "crap-patch-no-gui"?  ;)

@DSheirer  Is there a way that a headless option could be made available using this method?
Use case scenario would be use GUI to set up your system (CCs, tuners, aliases, TGIDs, broadcastify, etc), and then use a separate bash script/sdrtrunk.jar that runs GUI-less from the terminal.  Maybe list call events in the terminal too?  Thoughts?

**Comment by DSheirer** [comment]
A headless option is something we can do.

**Comment by ImagoTrigger** [comment]
a soundless option would be cool also.

**Comment by edbir1** [comment]
I am attempting to stream a P25 Phase1 system on a Raspberry Pi 3+ using 2 dongles.  It works, but I need to automagically start at boot, but I get the HeadlessException error which I think is complaining that I don't have a windows manager open.  I also noticed that if I close the gui window when I am running sdrtrunk from the desktop, the cpu usage drops to a usable level.  It is nice to have a trunking program which will run on a rpi!
1.  How can I run sdrtrunk at boot headless?

**Comment by DSheirer** [comment]
There currently isn't a way to run sdrtrunk headless.

It would require a new class that performs all of the same startup operations that the SDRTrunk.java class performs, but without instantiating the gui pieces.

**Comment by edbir1** [comment]
bummer.  
The ability to run headless and P25 Phase 2 are top on my Please list.
Thanks for writing the program and the fact that it will work on a rpi is awesome.

**Comment by ImagoTrigger** [comment]
i made a nogui nosound patch right around v.3.  thus prevented jvm from
using gfx and snd libs completley.  basically i just copypastad .gui to
gui2 and commente out some jframe stuff....real fugly but it worked and it
ws awesome on a single core

On Thursday, October 25, 2018, Denny Sheirer <notifications@github.com>
wrote:
> @obihann A CLI version and a smaller UI option both sound like great
features to add. You're more than welcome to help out with the coding if
you'd like to contribute.
>
> The CLI version might be fairly straight forward ... using the pieces
constructed in the SDRTrunk.java class, minus the gui pieces. It would have
to use the previously stored playlist (channels, aliases, auto-start)
initially. I'm completely revamping the way that aliasable values are
handled within the code, with an end goal of creating an interface to radio
reference to pull down networks and user/group info and also resolving a
number of issues with the current approach. I eventually intend to add
support for multiple playlists, so the CLI could be updated to accept an
optional playlist parameter down the road.
>
> A small UI that presents the minimum details would be neat. Are you
familiar with JavaFX? I want to transition the current UI to JavaFX, but
just haven't had the time yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci4.googleusercontent.com/proxy/0Qp9VLoXO1ar3ONKjGAp_3GNWiOrnb_sIRfV-dXQjxlYEZLgE0yGxvAD5XwTnyUiOU7mXaKggKP8-OJKvBx8erJn5ts6K9MFf9DgbE8b-3PGuuHIq-cD0dn_cLUGEya4oLY25JDrej6XgZPeB5gm7Fk416LtAQ=s0-d-e1-ft#https://github.com/notifications/beacon/ADLfRFkvud6JzoELQEm19nhcfv3TyVzTks5uoXTwgaJpZM4Ldija.gif>

**Comment by n2cr** [comment]
> 
> 
> A headless option is something we can do.

Is this still a possibility?  Is there a command line that would work with 0.4.0 or 0.5.0alpha and Java 13 that would work without the need for code modifications?
Using an Airspy and would like to make the startup into a service which doesn't require a GUI.

**Comment by DSheirer** [comment]
It's still a possibility, but I haven't had any time to dedicte to it yet.

**Comment by n2cr** [comment]
You need to add “-Djava.awt.headless=true" to the Java arguments.  In the start script,  add:

set SDR_TRUNK_OPTS="-Djava.awt.headless=true"

right under the “set JAVA_EXE” line.

Works like a charm and reduces the load on small systems.

**Comment by SethArchambault** [comment]
Thanks!
